src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs

```diff
@@ -9,90 +9,102 @@ namespace System.Runtime.InteropServices.Marshalling
     /// Marshaller for ANSI strings
     /// </summary>
     [CLSCompliant(false)]
-    [CustomTypeMarshaller(typeof(string), BufferSize = 0x100,
-        Features = CustomTypeMarshallerFeatures.UnmanagedResources | CustomTypeMarshallerFeatures.TwoStageMarshalling | CustomTypeMarshallerFeatures.CallerAllocatedBuffer)]
-    public unsafe ref struct AnsiStringMarshaller
+    [CustomMarshaller(typeof(string), MarshalMode.Default, typeof(AnsiStringMarshaller))]
+    [CustomMarshaller(typeof(string), MarshalMode.ManagedToUnmanagedIn, typeof(ManagedToUnmanagedIn))]
+    public static unsafe class AnsiStringMarshaller
     {
-        private byte* _nativeValue;
-        private bool _allocated;
+        /// <summary>
+        /// Convert a string to an unmanaged version.
+        /// </summary>
+        /// <param name="managed">A managed string</param>
+        /// <returns>An unmanaged string</returns>
+        public static byte* ConvertToUnmanaged(string? managed)
+        {
+            if (managed is null)
+                return null;
+
+            int exactByteCount = Marshal.GetAnsiStringByteCount(managed); // Includes null terminator
+            Span<byte> buffer = new ((byte*)NativeMemory.Alloc((nuint)exactByteCount), exactByteCount);
```

Comment:
This has to use CoTaskMemAlloc. It is part of the contract for marshalling strings as `ref` and `out`.